app/uk/gov/hmrc/pensionschemereturn/auth/PsrAuth.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pensionschemereturn.auth
+
+import play.api.Logging
+import play.api.mvc.{Request, Result}
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.{~, Name}
+import uk.gov.hmrc.auth.core.{AuthorisedFunctions, Enrolment, Enrolments}
+import uk.gov.hmrc.http.{BadRequestException, HeaderCarrier, UnauthorizedException}
+import uk.gov.hmrc.pensionschemereturn.config.Constants.{psaEnrolmentKey, psaIdKey, pspEnrolmentKey, pspIdKey}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+final case class PsrAuthContext[A](
+  externalId: String,
+  psaPspId: String,
+  name: Option[Name],
+  request: Request[A]
+)
+
+trait PsrAuth extends AuthorisedFunctions with Logging {
+
+  private val AuthPredicate = Enrolment(psaEnrolmentKey).or(Enrolment(pspEnrolmentKey))
+  private val PsrRetrievals = Retrievals.externalId.and(Retrievals.allEnrolments).and(Retrievals.name)
+
+  private type PsrAction[A] = PsrAuthContext[A] => Future[Result]
+
+  def authorisedAsPsrUser(
```

Comment:
Do we need some kind of test for this method?

Reply:
I will add unit tests for authentication

Reply:
tests are added